### DISCUSSION TITLE
foxglove websocket in C++ and JSON)

### DISCUSSION THREAD

**Discussion by GianlucaCerilli** [foxglove/community#983] [Q&A] [answered]
Good morning,

I was looking at [these examples](https://github.com/foxglove/ws-protocol/tree/main/cpp/examples/src) and was wondering if you also have a working example server in C++ that uses JSON messages.

In addition to this, I also wanted to ask you if it is possible to link _foxglove-websocket_ from CMakeLists in C++, without using Python and Conan.

Thanks

**Comment by achim-k** [comment, marked as answer]
https://github.com/foxglove/ws-protocol/pull/724 adds an example JSON C++ server

Yes, you can use foxglove-websocket directly with CMakelist, you probably have to use [`add_subdirectory()`](https://cmake.org/cmake/help/latest/command/add_subdirectory.html) or something similar.

**Reply by GianlucaCerilli** [reply]
Thanks!
Which parameters should be changed to publish two different example_server in parallel?
I would like to run two different executables and see their objects published on the same Foxglove 3D panel (so I think I should use the same port for connection).

Thanks a lot

**Reply by achim-k** [reply]
Foxglove can only connect to one websocket server at a time. However, you can publish two different topics from the same server and visualize both topics in the 3D panel.

**Reply by GianlucaCerilli** [reply]
I see. Is it possible to establish a connection to one single server and publish topics from two different executables? They should both use the same server somehow.
So, for example:
* Executable 1 - It just creates a server for connecting to Foxglove
* Executable 2 - It uses that server to publish a cube (for example)
* Executable 3 - It uses that server to publish a sphere (for example)

Or a similar solution that could allow to publish different topics in parallel, from different executables though.

Thanks!

**Reply by jtbandes** [reply]
No, since Foxglove only connects to one server at a time, you would have to manage the inter-process communication yourself so the messages can be published from executable 1. (This is the role played by DDS, TCPROS, or other middlewares.)

**Reply by GianlucaCerilli** [reply]
Ok, thanks...is there a way to add the channels dynamically, instead?

For example, I have a _run_ function that fills a `std::vector<foxglove::ChannelWithoutId> channels` iteratively, as soon as it discovers new topics.
Then, at each cycle the `server->addChannels` is called, taking the `channels` as parameter.

I can see the list of topics on Foxglove, but without content and with this error:

_Multiple channels advertise on the same topic: **topic_name** (**number1** and **number2**)_.

_number2_ seems to be the ChannelId, and I think _number1_ is the handler id.
I also notice _number2_ always increases, is there a way to reset it at each iteration? I guess calling `server->removeChannels` soon before the `server->addChannels` is not the solution.

Thanks a lot

**Reply by jtbandes** [reply]
It’s ok to add channels dynamically. Just call addChannels with only the new ones.
